src/librustc_typeck/outlives/infer.rs

```diff
@@ -0,0 +1,79 @@
+// Copyright 2013 The Rust Project Developers. See the COPYRIGHT
+// file at the top-level directory of this distribution and at
+// http://rust-lang.org/COPYRIGHT.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+
+type DefIdPredicatesMap = <DefId, Vec<OutlivePredicate>>
+pub struct OutlivePredicate {}
+
+pub struct OutlivesCrateVisitor<'tcx> {
+    tcx: TyCtxt<'tcx, 'tcx, 'tcx>,
+    predicates: FxHashMap<_>,
+    empty_predicate: Rc<CratePredicatesMap<'tcx>>,
+}
+
+impl OutlivesCrateVisitor {
+
+    fn add_outlives_for_item(&mut self, id: ast::NodeId, outlive_map: DefIdPredicatesMap) {
+        //we will add things to outlives_map
+
+        let predicates = self.tcx.explicit_predicates_of(def_id);
+        let filtered: Vec<_> = predicates
+            .predicates.iter()
+            .filter(|pred|
+                match *pred {
+                    &ty::Predicate::TypeOutlives(..) | &ty::Predicate::RegionOutlives(..) => true,
+                    _ => false,
+                }
+            )
+            .collect::<Vec<_>>();
+
+        //now infer outlive prediacates from the filtered predicates
+        for p in filtered {
+            the_great_inferring(p)
+        }
```

Comment:
This doesn't look quite right. This is taking a single item (`def_id`) -- e.g., a single struct or enum type -- and computing its explicit predicates, and then iterating over those predicates. What we want to do be doing is something [more like this](https://github.com/rust-lang/rust/issues/44493#issuecomment-343519092).